### PR TITLE
Error stack safety

### PR DESCRIPTION
**Pull request by rachel-bousfield**
# Error Recovery Policy
This PR introduces the ability for the replay machine to enable & disable the error guard stack via the new `SetErrorPolicy` WAVM opcode.

When disabled, errors are unguarded, inducing a chain halt on failure. Though counter-intuitive, this ability turns certain theoretical classes of attacks into chain halts rather than arbitrary code execution in the STF -- not that any such attacks are known.

# Merged Host I/O Implementations
Additionally, this PR reduces both sets of host I/O definitions into a single implementation via the new `UserHost` trait. This ensures consistent behavior across all three scenarios: native, proving, and jit. The two `host.rs` files still exist, but each points to the trait in the new `user-host-trait` crate.

# Auto-generated `forward` and `forward_stub`.
The new `forward` crate generates `forward.wat` and `forward_stub.wat` for us. Beyond general safety, this allows us to do two things
1. Reserve the first 512 function offsets so that more Host I/O ops may be added without requiring re-activation
2. Introduce a new `set-trap` affordance so that `user-host` can induce recovery in the forwarder

Point 2 is important in that it moves error recovery from `user-host` to the forwarder, ensuring any un-recovered pieces of state are accounted for (in this case just a single global variable, the trap flag). Though probably not necessary, this should eliminate an entire class of theoretical attacks on `user-host` and the chain.

# Miscellaneous
Other improvements have been made to the types we use when referring to byte counts in pricing math and stack pointer advancement.

Protecting `user-host` required adding a new internal func, `UserMemorySize`, which allows `user-host` to check the safety of memory offsets before accessing contents that are out-of-bounds. It's worth reviewing that `user-host` is free of incidental panics now.

Associated PRs
- https://github.com/OffchainLabs/stylus-contracts/pull/24

### CODE REVIEW

File: arbitrator/wasm-libraries/user-host/src/host.rs [hunk 1]
```diff
@@ -1,331 +1,690 @@
...
 
-#[no_mangle]
-pub unsafe extern "C" fn user_host__write_result(ptr: usize, len: usize) {
-    let program = Program::start(0);
-    program.pay_for_read(len as u64).unwrap();
-    program.outs = wavm::read_slice_usize(ptr, len);
+macro_rules! trace {
+    ($name:expr, $env:expr, [$($args:expr),+], [$($outs:expr),+], $ret:expr) => {{
+        if $env.evm_data().tracing {
+            //let start_ink = $env.start_ink;
```
Comment (rauljordan) [inline arbitrator/wasm-libraries/user-host/src/host.rs]:
```suggestion
```

File: arbitrator/wasm-libraries/user-host/src/host.rs [hunk 2]
```diff
@@ -1,331 +1,690 @@
...
+
+    fn write_result(&mut self, ptr: u32, len: u32) -> Result<(), Self::Err> {
+        self.buy_ink(HOSTIO_INK)?;
+        self.pay_for_read(len.into())?;
+        *self.outs() = self.read_slice(ptr, len)?;
+        trace!("write_result", self, &*self.outs(), &[])
+    }
+
+    fn storage_load_bytes32(&mut self, key: u32, dest: u32) -> Result<(), Self::Err> {
+        self.buy_ink(HOSTIO_INK + 2 * PTR_INK + EVM_API_INK)?;
```
Comment (rauljordan) [inline arbitrator/wasm-libraries/user-host/src/host.rs]: Comment could be helpful to explain the 2 magic number

File: arbitrator/wasm-libraries/user-host/src/host.rs [hunk 3]
```diff
@@ -1,331 +1,690 @@
...
+    fn write_bytes20(&self, ptr: u32, src: Bytes20) -> Result<(), Self::MemoryErr>;
+    fn write_bytes32(&self, ptr: u32, src: Bytes32) -> Result<(), Self::MemoryErr>;
+    fn write_slice(&self, ptr: u32, src: &[u8]) -> Result<(), Self::MemoryErr>;
+
+    fn say<D: Display>(&self, text: D);
+    fn trace(&self, name: &str, args: &[u8], outs: &[u8], end_ink: u64);
+
+    fn read_args(&mut self, ptr: u32) -> Result<(), Self::Err> {
+        self.buy_ink(HOSTIO_INK)?;
+        self.pay_for_write(self.args().len() as u64)?;
```
Comment (rauljordan) [inline arbitrator/wasm-libraries/user-host/src/host.rs]: Curious why pay for write takes in a u64...could we ever pay for that amount of data?

Comment (rachel-bousfield) [reply arbitrator/wasm-libraries/user-host/src/host.rs]: Ink is a `u64`, which historically made the payment functions easier to work with. 

Looking at our new impl though, it's def cleaner to use `u32` everywhere. I went ahead and changed these, nice simplification :)

File: arbitrator/wasm-libraries/user-host/src/host.rs (hunk 2 above)
Comment (rachel-bousfield) [reply arbitrator/wasm-libraries/user-host/src/host.rs]: The `2` comes from the number of pointer ops. There's a `read_bytes32` and a `write_bytes32`, so we charge for two `PTR_INK`'s

**Review by joshuacolvin0 — APPROVED** [review]

**Review by rauljordan — APPROVED** [review]

**Review by joshuacolvin0 — APPROVED** [review]
LGTM

**Pull request merged**